File: 2015/d06/d06.nim
from std/os import commandLineParams
import std/re
import std/sequtils
import std/strformat
import std/strutils

type
  LightGrid[T] = array[1_000, array[1_000, T]]
  Action = enum on, off, toggle
  Pos = tuple[x: int, y: int]
  Command = object
    action: Action
    fromPos: Pos
    toPos: Pos

proc parseCommand(s: string): Command =
  let pat = re"(toggle|turn off|turn on) (\d+),(\d+) through (\d+),(\d+)"
  var groups: array[5, string]
  let i = find(s, pat, groups)
  if i >= 0:
    var a = case groups[0]
      of "toggle": Action.toggle
      of "turn off": Action.off
      of "turn on": Action.on
      else: raise newException(Exception, "Invalid action")
    return Command(
      action: a,
      fromPos: (parseInt(groups[1]), parseInt(groups[2])),
      toPos: (parseInt(groups[3]), parseInt(groups[4])),
    )
  else:
    raise newException(Exception, fmt"Couldn't parse command '{s}'")

proc applyCommand(grid: var LightGrid[bool], cmd: Command): void =
  for x in cmd.fromPos.x..cmd.toPos.x:
    for y in cmd.fromPos.y..cmd.toPos.y:
      case cmd.action
      of Action.toggle: grid[x][y] = not grid[x][y]
      of Action.on: grid[x][y] = true
      of Action.off: grid[x][y] = false

proc applyCommand(grid: var LightGrid[uint], cmd: Command): void =
  for x in cmd.fromPos.x..cmd.toPos.x:
    for y in cmd.fromPos.y..cmd.toPos.y:
      case cmd.action
      of Action.toggle: grid[x][y] = grid[x][y] + 2
      of Action.on: grid[x][y] = grid[x][y] + 1
      of Action.off: grid[x][y] = grid[x][y] - min(grid[x][y], 1)

proc applyAll[T](grid: var LightGrid[T], cmds: seq[Command]): void =
  for c in cmds:
    grid.applyCommand(c)

func litCount(grid: LightGrid[bool]): int =
  result = 0
  for x in 0..999:
    for y in 0..999:
      if grid[x][y]:
        result += 1
  return result

func totalBrightness(grid: LightGrid[uint]): uint =
  result = 0
  for x in 0..999:
    for y in 0..999:
      result += grid[x][y]
  return result

proc main(): void =
  let inputFilename = commandLineParams()[0]
  let lines = toSeq(inputFilename.lines)
  let commands = lines.map(parseCommand)

  var grid: LightGrid[bool]
  grid.applyAll(commands)
  echo "p1: there are ", grid.litCount(), " lights on"

  var grid2: LightGrid[uint]
  grid2.applyAll(commands)
  echo "p2: total brightness is ", grid2.totalBrightness()

main()

File: 2015/d06/d06.py
#!/usr/bin/env python

from sys import argv
from enum import Enum
import re

class Action(Enum):
    TOGGLE = "toggle"
    TURN_ON = "turn on"
    TURN_OFF = "turn off"

class Pos:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Command:
    def __init__(self, action, from_pos, to_pos):
        self.action = action
        self.from_pos = from_pos
        self.to_pos = to_pos

PAT = re.compile("(toggle|turn off|turn on) (\d+),(\d+) through (\d+),(\d+)")

def new_grid(init_val):
    grid = []
    for i in range(0, 1_000):
        grid.append([init_val] * 1_000)
    return grid

def parse_cmd(cmd_str):
    m = PAT.match(cmd_str)
    if not m:
        raise Exception(f"pattern didn't match '{cmd_str}'")
    if m[1] == "toggle":
        act = Action.TOGGLE
    elif m[1] == "turn on":
        act = Action.TURN_ON
    elif m[1] == "turn off":
        act = Action.TURN_OFF
    else:
        raise Exception("invalid action from command")
    return Command(
            act,
            Pos(int(m[2]), int(m[3])),
            Pos(int(m[4]), int(m[5])),
            )

def apply_p1(grid, action, x, y):
    if action == Action.TOGGLE:
        grid[x][y] = not grid[x][y]
    elif action == Action.TURN_ON:
        grid[x][y] = True
    elif action == Action.TURN_OFF:
        grid[x][y] = False
    else:
        raise Exception("invalid action in apply_p1")

def apply_p2(grid, action, x, y):
    if action == Action.TOGGLE:
        grid[x][y] = grid[x][y] + 2
    elif action == Action.TURN_ON:
        grid[x][y] = grid[x][y] + 1
    elif action == Action.TURN_OFF:
        grid[x][y] = max(0, grid[x][y] - 1)
    else:
        raise Exception("invalid action in apply_p2")

def apply_range(grid, cmd, fn):
  for x in range(cmd.from_pos.x, cmd.to_pos.x + 1):
    for y in range(cmd.from_pos.y, cmd.to_pos.y + 1):
        fn(grid, cmd.action, x, y)

def apply_all(grid, commands, fn):
    for c in commands:
        apply_range(grid, c, fn)

def count_lit(grid):
    cols_sums = [
        sum([1 for cell in col if cell == True])
        for col in grid
        ]
    return sum(cols_sums)

def total_brightness(grid):
    cols_sums = [sum(col) for col in grid]
    return sum(cols_sums)

with open(argv[1], "r") as fh:
    cmds = [parse_cmd(l) for l in fh.readlines()]
    grid1 = new_grid(False)
    apply_all(grid1, cmds, apply_p1)
    print(f"p1: {count_lit(grid1)} lights are on")

    grid2 = new_grid(0)
    apply_all(grid2, cmds, apply_p2)
    print(f"p2: total brightness is {total_brightness(grid2)}")
